File: packages/go/storage/sqlite/sqlite.go
package sqlite

import (
	"context"
	"database/sql"
	"encoding/hex"
	"errors"
	"fmt"
	"regexp"
	"strings"

	identifold "github.com/greyfoundry/identifold/packages/go"
	"github.com/greyfoundry/identifold/packages/go/storage"
)

var sequentialReference = regexp.MustCompile(`^([A-Z]{2,8})-(?:([0-9]{4})-)?([0-9]{4,18})-[0-9A-Z*~$=U]$`)

// Adapter uses a caller-owned database/sql handle backed by SQLite.
type Adapter struct {
	database *sql.DB
}

func New(database *sql.DB) *Adapter {
	return &Adapter{database: database}
}

func (adapter *Adapter) Reserve(ctx context.Context, request storage.ReferenceReservation) (bool, error) {
	machineID, err := machineIDBytes(request.MachineID)
	if err != nil {
		return false, allocationError("allocation_conflict")
	}
	result, err := adapter.database.ExecContext(ctx,
		`INSERT INTO identifold_references (reference, namespace, machine_id)
		 VALUES (?, ?, ?) ON CONFLICT(reference) DO NOTHING`,
		request.Reference, request.Namespace, machineID,
	)
	if err != nil {
		return false, allocationError("allocation_conflict")
	}
	changes, err := result.RowsAffected()
	if err != nil {
		return false, allocationError("allocation_conflict")
	}
	return changes == 1, nil
}

func (adapter *Adapter) Resolve(ctx context.Context, reference string, namespace string) (*storage.ReferenceMapping, error) {
	var machineID []byte
	var resolvedNamespace string
	err := adapter.database.QueryRowContext(ctx,
		`SELECT machine_id, namespace FROM identifold_references
		 WHERE reference = ? AND namespace = ?`, reference, namespace,
	).Scan(&machineID, &resolvedNamespace)
	if errors.Is(err, sql.ErrNoRows) {
		parts := sequentialReference.FindStringSubmatch(reference)
		if parts == nil {
			return nil, nil
		}
		err = adapter.database.QueryRowContext(ctx,
			`SELECT machine_id, namespace FROM identifold_sequence_allocations
			 WHERE namespace = ? AND reference_prefix = ? AND scope = ? AND sequence = ?`,
			namespace, parts[1], parts[2], parts[3],
		).Scan(&machineID, &resolvedNamespace)
	}
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, allocationError("allocation_conflict")
	}
	value, err := bytesMachineID(machineID)
	if err != nil {
		return nil, allocationError("allocation_conflict")
	}
	return &storage.ReferenceMapping{MachineID: value, Namespace: resolvedNamespace}, nil
}

func (adapter *Adapter) Allocate(ctx context.Context, request storage.SequenceAllocationRequest) (uint64, error) {
	if request.Width < 4 || request.Width > 18 {
		return 0, allocationError("invalid_allocation_policy")
	}
	machineID, err := machineIDBytes(request.MachineID)
	if err != nil {
		return 0, allocationError("allocation_conflict")
	}
	scope := ""
	if request.Scope != nil {
		scope = *request.Scope
	}
	connection, err := adapter.database.Conn(ctx)
	if err != nil {
		return 0, allocationError("allocation_conflict")
	}
	defer connection.Close()
	if _, err = connection.ExecContext(ctx, "BEGIN IMMEDIATE"); err != nil {
		return 0, allocationError("allocation_conflict")
	}
	committed := false
	defer func() {
		if !committed {
			_, _ = connection.ExecContext(context.Background(), "ROLLBACK")
		}
	}()

	var existing int64
	var existingPrefix string
	var existingWidth uint8
	err = connection.QueryRowContext(ctx,
		`SELECT sequence, reference_prefix, width
		 FROM identifold_sequence_allocations
		 WHERE namespace = ? AND scope = ? AND machine_id = ?`,
		request.Namespace, scope, machineID,
	).Scan(&existing, &existingPrefix, &existingWidth)
	if err == nil {
		if existingPrefix != request.ReferencePrefix || existingWidth != request.Width {
			return 0, allocationError("invalid_allocation_policy")
		}
		if _, err = connection.ExecContext(ctx, "COMMIT"); err != nil {
			return 0, allocationError("allocation_conflict")
		}
		committed = true
		return uint64(existing), nil
	}
	if !errors.Is(err, sql.ErrNoRows) {
		return 0, allocationError("allocation_conflict")
	}
	_, err = connection.ExecContext(ctx,
		`INSERT INTO identifold_sequences
		 (namespace, scope, reference_prefix, width, last_value)
		 VALUES (?, ?, ?, ?, 0) ON CONFLICT DO NOTHING`,
		request.Namespace, scope, request.ReferencePrefix, request.Width,
	)
	if err != nil {
		return 0, allocationError("allocation_conflict")
	}
	var prefix string
	var width uint8
	var current int64
	err = connection.QueryRowContext(ctx,
		`SELECT reference_prefix, width, last_value FROM identifold_sequences
		 WHERE namespace = ? AND scope = ?`, request.Namespace, scope,
	).Scan(&prefix, &width, &current)
	if err != nil {
		return 0, allocationError("allocation_conflict")
	}
	if prefix != request.ReferencePrefix || width != request.Width {
		return 0, allocationError("invalid_allocation_policy")
	}
	maximum := int64(1)
	for range request.Width {
		maximum *= 10
	}
	maximum--
	if current >= maximum {
		return 0, allocationError("sequence_overflow")
	}
	allocated := current + 1
	_, err = connection.ExecContext(ctx,
		`UPDATE identifold_sequences SET last_value = ?
		 WHERE namespace = ? AND scope = ?`, allocated, request.Namespace, scope,
	)
	if err == nil {
		_, err = connection.ExecContext(ctx,
			`INSERT INTO identifold_sequence_allocations
			 (namespace, scope, sequence, machine_id, reference_prefix, width)
			 VALUES (?, ?, ?, ?, ?, ?)`,
			request.Namespace, scope, allocated, machineID, request.ReferencePrefix, request.Width,
		)
	}
	if err != nil {
		return 0, allocationError("allocation_conflict")
	}
	if _, err = connection.ExecContext(ctx, "COMMIT"); err != nil {
		return 0, allocationError("allocation_conflict")
	}
	committed = true
	return uint64(allocated), nil
}

func machineIDBytes(value string) ([]byte, error) {
	decoded, err := hex.DecodeString(strings.ReplaceAll(value, "-", ""))
	if err != nil || len(decoded) != 16 {
		return nil, errors.New("invalid machine identifier")
	}
	return decoded, nil
}

func bytesMachineID(value []byte) (string, error) {
	if len(value) != 16 {
		return "", errors.New("invalid machine identifier")
	}
	hexValue := hex.EncodeToString(value)
	return fmt.Sprintf("%s-%s-%s-%s-%s", hexValue[:8], hexValue[8:12], hexValue[12:16], hexValue[16:20], hexValue[20:]), nil
}

func allocationError(code string) error {
	return &identifold.Error{Code: code}
}

var _ storage.Adapter = (*Adapter)(nil)

File: packages/typescript/src/sqlite.ts
import { IdentifoldError } from "./errors.js";
import {
  bytesToMachineId,
  machineIdToBytes,
  parseSequentialReference,
} from "./sql-storage.js";
import { createStorageAdapter } from "./storage.js";
import type { ReferenceLookup } from "./resolver.js";
import type {
  ReferenceReservation,
  ReferenceStore,
  SequenceAllocationRequest,
  SequenceAllocator,
} from "./service.js";
import type { IdentifoldStorageAdapter } from "./storage.js";

interface SqliteRunResult {
  readonly changes: number;
}

interface SqliteStatement {
  get(...values: unknown[]): unknown;
  run(...values: unknown[]): SqliteRunResult;
}

export interface SqliteDatabase {
  exec(text: string): unknown;
  prepare(text: string): SqliteStatement;
}

export function createSqliteReferenceStore(
  database: SqliteDatabase,
): ReferenceStore {
  const statement = database.prepare(
    `INSERT INTO identifold_references (reference, namespace, machine_id)
     VALUES (?, ?, ?)
     ON CONFLICT(reference) DO NOTHING`,
  );
  return Object.freeze({
    async reserve(reservation: ReferenceReservation) {
      await Promise.resolve();
      try {
        return (
          statement.run(
            reservation.reference,
            reservation.namespace,
            machineIdToBytes(reservation.machineId),
          ).changes === 1
        );
      } catch (error) {
        throw mapSqliteError(error);
      }
    },
  });
}

export function createSqliteSequenceAllocator(
  database: SqliteDatabase,
): SequenceAllocator {
  return Object.freeze({
    async allocate(request: SequenceAllocationRequest) {
      await Promise.resolve();
      if (
        !Number.isInteger(request.width) ||
        request.width < 4 ||
        request.width > 18
      ) {
        throw new IdentifoldError(
          "invalid_allocation_policy",
          "Sequential allocation policy is invalid",
        );
      }
      const scope = request.scope ?? "";
      try {
        database.exec("BEGIN IMMEDIATE");
        const existing = recordOrNull(
          database
            .prepare(
              `SELECT CAST(sequence AS TEXT) AS sequence, reference_prefix, width
               FROM identifold_sequence_allocations
               WHERE namespace = ? AND scope = ? AND machine_id = ?`,
            )
            .get(request.namespace, scope, machineIdToBytes(request.machineId)),
        );
        if (existing !== null) {
          if (
            existing.reference_prefix !== request.referencePrefix ||
            existing.width !== request.width ||
            typeof existing.sequence !== "string"
          ) {
            throw new IdentifoldError(
              "invalid_allocation_policy",
              "Sequential allocation policy is invalid",
            );
          }
          database.exec("COMMIT");
          return BigInt(existing.sequence);
        }

        database
          .prepare(
            `INSERT INTO identifold_sequences
               (namespace, scope, reference_prefix, width, last_value)
             VALUES (?, ?, ?, ?, 0)
             ON CONFLICT DO NOTHING`,
          )
          .run(
            request.namespace,
            scope,
            request.referencePrefix,
            request.width,
          );
        const state = recordOrNull(
          database
            .prepare(
              `SELECT reference_prefix, width, CAST(last_value AS TEXT) AS last_value
               FROM identifold_sequences
               WHERE namespace = ? AND scope = ?`,
            )
            .get(request.namespace, scope),
        );
        if (state === null) {
          throw new IdentifoldError(
            "invalid_allocation_policy",
            "Sequential allocation policy is invalid",
          );
        }
        if (
          state.reference_prefix !== request.referencePrefix ||
          state.width !== request.width ||
          typeof state.last_value !== "string"
        ) {
          throw new IdentifoldError(
            "invalid_allocation_policy",
            "Sequential allocation policy is invalid",
          );
        }
        const current = BigInt(state.last_value);
        const maximum = 10n ** BigInt(request.width) - 1n;
        if (current >= maximum) {
          throw new IdentifoldError(
            "sequence_overflow",
            "Sequential reference capacity is exhausted",
          );
        }
        const allocated = current + 1n;
        database
          .prepare(
            `UPDATE identifold_sequences SET last_value = ?
             WHERE namespace = ? AND scope = ?`,
          )
          .run(allocated, request.namespace, scope);
        database
          .prepare(
            `INSERT INTO identifold_sequence_allocations
               (namespace, scope, sequence, machine_id, reference_prefix, width)
             VALUES (?, ?, ?, ?, ?, ?)`,
          )
          .run(
            request.namespace,
            scope,
            allocated,
            machineIdToBytes(request.machineId),
            request.referencePrefix,
            request.width,
          );
        database.exec("COMMIT");
        return allocated;
      } catch (error) {
        try {
          database.exec("ROLLBACK");
        } catch {
          // The transaction may already have completed.
        }
        throw mapSqliteError(error);
      }
    },
  });
}

export function createSqliteReferenceLookup(
  database: SqliteDatabase,
): ReferenceLookup {
  return (reference, namespace) =>
    Promise.resolve().then(() => {
      try {
        const random = recordOrNull(
          database
            .prepare(
              `SELECT machine_id, namespace FROM identifold_references
             WHERE reference = ? AND namespace = ?`,
            )
            .get(reference, namespace),
        );
        if (random !== null) return mappingFromRecord(random);

        const parts = parseSequentialReference(reference);
        if (parts === null) return null;
        const sequential = recordOrNull(
          database
            .prepare(
              `SELECT machine_id, namespace
             FROM identifold_sequence_allocations
             WHERE namespace = ? AND reference_prefix = ?
               AND scope = ? AND sequence = ?`,
            )
            .get(namespace, parts.prefix, parts.scope, parts.sequence),
        );
        return sequential === null ? null : mappingFromRecord(sequential);
      } catch (error) {
        throw mapSqliteError(error);
      }
    });
}

export function createSqliteStorageAdapter(
  database: SqliteDatabase,
): IdentifoldStorageAdapter {
  return createStorageAdapter({
    referenceStore: createSqliteReferenceStore(database),
    sequenceAllocator: createSqliteSequenceAllocator(database),
    lookup: createSqliteReferenceLookup(database),
  });
}

function recordOrNull(value: unknown): Record<string, unknown> | null {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function mappingFromRecord(record: Record<string, unknown>) {
  if (typeof record.namespace !== "string") {
    throw new IdentifoldError(
      "allocation_conflict",
      "Reference lookup returned an invalid result",
    );
  }
  return Object.freeze({
    machineId: bytesToMachineId(record.machine_id),
    namespace: record.namespace,
  });
}

function mapSqliteError(error: unknown): IdentifoldError {
  if (error instanceof IdentifoldError) return error;
  return new IdentifoldError(
    "allocation_conflict",
    "Database allocation could not be committed",
  );
}
